xblockutils/studio_editable.py

```diff
@@ -0,0 +1,263 @@
+        for field_name in self.editable_fields:
+            field = self.fields[field_name]
+            assert field.scope in (Scope.content, Scope.settings)
+            field_info = self._make_field_info(field_name, field)
+            if field_info is not None:
+                context["fields"].append(field_info)
+        fragment.content = loader.render_template('templates/studio_edit.html', context)
+        fragment.add_javascript(loader.load_unicode('public/studio_edit.js'))
+        fragment.initialize_js('StudioEditableXBlockMixin')
+        return fragment
+
+    def _make_field_info(self, field_name, field):
+        """
+        Create the information that the template needs to render a form field for this field.
+        """
+        supported_field_types = (
+            (Integer, 'integer'),
+            (Float, 'float'),
+            (Boolean, 'boolean'),
+            (String, 'string'),
+            (JSONField, 'generic'),  # This is last so as a last resort we display a text field w/ the JSON string
+        )
+        info = {
+            'name': field_name,
+            'display_name': field.display_name,
+            'is_set': field.is_set_on(self),
+            'default': field.default,
+            'value': field.read_from(self),
+            'has_values': False,
+            'help': field.help,
+            'allow_reset': field.runtime_options.get('resettable_editor', True),
+        }
+        for type_class, type_name in supported_field_types:
+            if isinstance(field, type_class):
+                info['type'] = type_name
+                # If String fields are declared like String(..., multiline_editor=True), then call them "text" type:
+                editor_type = field.runtime_options.get('multiline_editor')
+                if type_class is String and editor_type:
+                    if editor_type == "html":
+                        info['type'] = 'html'
+                    else:
+                        info['type'] = 'text'
+                break
+        if "type" not in info:
+            return None  # This is not a field we can support (it doesn't even have a JSON representation)
+        if info["type"] == "generic":
+            # Convert value to JSON string if we're treating this field generically:
+            info["value"] = json.dumps(info["value"])
+            info["default"] = json.dumps(info["default"])
+        if field.values and not isinstance(field, Boolean):
+            info['has_values'] = True
+            # This field has only a limited number of pre-defined options.
+            # Protip: when defining the field, values= can be a callable.
+            if isinstance(field.values, dict) and isinstance(field, (Float, Integer)):
+                # e.g. {"min": 0 , "max": 10, "step": .1}
+                info["min"] = field.values["min"]
+                info["max"] = field.values["max"]
+                info["step"] = field.values["step"]
```

Comment:
No, there shouldn't be. But if for some reason a key like "default" were to be set in field.values, it would mess up our info dict if we are using info.update(field.values). I think I prefer the more explicit approach in this case, so that it's clear (A) what's going on here and (B) what fields are expected/allowed.